### PR TITLE
Support Windows x64

### LINKED ISSUE #11: How to use in Windows DLL (64-bit)?
I'm using this [simple example](https://hackage.haskell.org/package/pipes-http-1.0.2/docs/Pipes-HTTP.html) from `Pipes.HTTP`

```Haskell
import Pipes
import Pipes.HTTP
import qualified Pipes.ByteString as PB  -- from `pipes-bytestring`

printPage = do
    req <- parseUrl "https://www.example.com"
    withManager tlsManagerSettings $ \m ->
        withHTTP req m $ \resp ->
            runEffect $ responseBody resp >-> PB.stdout
```

From within GHCi, this function works fine. But when I call it through the FFI in my C++ application, I get "`user error: cryptReleaseCtx`" printed to the screen. It appears that this error is raised [here](https://github.com/vincenthz/hs-crypto-random/blob/master/Crypto/Random/Entropy/Windows.hs#L82).

Do you have any thoughts on how I can track this down further? Is there some static initialization that is getting missed when running through the FFI? I'll try a 32-bit DLL next.

### PR DESCRIPTION
**Pull request by 3noch**
Fixes #11

### CODE REVIEW

**Comment by 3noch** [comment]
@vincenthz I didn't run any official tests, other than a simple test app that used to crash on x64, but now does not.

**Comment by 3noch** [comment]
@vincenthz I cleaned up the code a bit from a consistency perspective and ran one of the tests. I'm not sure how to verify that the test passed other than the fact that it produced a large file with random data. That's the point, I think, no?

File: Crypto/Random/Entropy/Windows.hs [hunk 1]
```diff
@@ -43,26 +35,52 @@ instance EntropySource WinCryptoAPI where
     entropyGather WinCryptoAPI ptr n = do
         mctx <- cryptAcquireCtx
         case mctx of
-            Nothing  -> error "cannot re-grab win crypto api"
+            Nothing  -> do
+                lastError <- getLastError
+                fail $ "cannot re-grab win crypto api: error " ++ show lastError
```
Comment (3noch) [inline Crypto/Random/Entropy/Windows.hs]: This used to use `error`, but I changed it to `fail` because `cryptAcquireCtx` was using `fail`.

**Comment by vincenthz** [comment]
Thanks, presumably it was the CryptCtx that was not supposed to be a Word32, but a 32/64bits architecture dependent type ? is there some better windows types representing this better than using some C preprocessor on Word32/Word64 maybe ?

**Comment by 3noch** [comment]
Thanks for merging. I'm happy to make any further edits you'd like to see.

`CryptCtx` is an alias for `HCRYPTPROV` which is just a `typedef` for `ULONG_PTR`. The clip below is from `basetsd.h` where `ULONG_PTR` is defined differently for 32-bit and 64-bit.

``` C++
#if defined(_WIN64)
    typedef __int64 INT_PTR, *PINT_PTR;
    typedef unsigned __int64 UINT_PTR, *PUINT_PTR;

    typedef __int64 LONG_PTR, *PLONG_PTR;
    typedef unsigned __int64 ULONG_PTR, *PULONG_PTR;

    #define __int3264   __int64

#else
    typedef _W64 int INT_PTR, *PINT_PTR;
    typedef _W64 unsigned int UINT_PTR, *PUINT_PTR;

    typedef _W64 long LONG_PTR, *PLONG_PTR;
    typedef _W64 unsigned long ULONG_PTR, *PULONG_PTR;

    #define __int3264   __int32

#endif
```

I didn't do much research to see if there was a type--e.g. in `Win32`--that already switches between address widths. Some of what I did was a direct copy from the `Win32` package though.

**Comment by 3noch** [comment]
@vincenthz Anything I can do to help get this released to Hackage?

**Comment by vincenthz** [comment]
I guess my question was, whether or not, `Win32` is already having a type representing a 32/64 bits value depending on the architecture. I had a quick look around, and it doesn't seems to be the case. With that in mind, I can release to hackage

**Comment by 3noch** [comment]
There are _some_ overlapping types [here](https://github.com/haskell/win32/blob/master/System/Win32/Types.hs). I could use those for `BOOL`, etc. but I don't see anything defined differently depending on address model.

**Comment by 3noch** [comment]
It looks like `Win32` could use some improvements in this area as well. I will start looking into a PR for `Win32` but that might be a while before you can depend on such improvements.

**Comment by 3noch** [comment]
@vincenthz Do you think this could get pushed to hackage?

**Pull request merged**